### ISSUE TITLE
Selective material model output

### ISSUE THREAD

**Issue by gassmoeller** [geodynamics/aspect#2045]
At some point we need a way to tell the material model which properties are required for the current call to `evaluate`. This would make calls to expensive material models like the one in #2042 much cheaper, if the expensive properties do not need to be computed. Maybe another (optional) parameter to evaluate, similar to `UpdateFlags` of deal.II?

**Comment by tjhei** [comment]
Makes sense. We should be careful to be a) easy for the user (especially the "I don't care about performance and want to evaluate all properties" case) and b) backwards compatible. I think a flag makes sense.

If you add an additional parameter to evaluate (``evaluate(in,out,flags)``), every material model needs to be adapted or we need a new evaluate function with the additional argument with a default implementation that calls the current one. Having two evaluate functions in the material model is not ideal or what do you think?

Another option would be to add a flag inside the material model inputs.

How do we want to handle places that call ``evaluate()``? I don't think that happens a lot in user code, so we could go through all places and change them immediately. In the long term I would like that the compiler enforces that you specify the correct flags (warnings or errors). I guess we could deprecate the old ``evaluate`` or force you to specify the flags in the constructor of the model inputs.
